File: go/bcenvelope/cbor.go
package bcenvelope

import (
	"fmt"

	bccomponents "github.com/nickel-blockchaincommons/bccomponents-go"
	bctags "github.com/nickel-blockchaincommons/bctags-go"
	dcbor "github.com/nickel-blockchaincommons/dcbor-go"
	knownvalues "github.com/nickel-blockchaincommons/knownvalues-go"
)

// Envelope CBOR tags.
var envelopeCBORTags = dcbor.TagsForValues([]dcbor.TagValue{bctags.TagEnvelope})

// EnvelopeCBORTags returns the CBOR tags used for Envelope.
func EnvelopeCBORTags() []dcbor.Tag {
	return envelopeCBORTags
}

// CBORTags returns the CBOR tags for this envelope.
func (e *Envelope) CBORTags() []dcbor.Tag {
	return envelopeCBORTags
}

// UntaggedCBOR returns the untagged CBOR representation of the envelope.
func (e *Envelope) UntaggedCBOR() dcbor.CBOR {
	switch c := e.envelopeCase.(type) {
	case *NodeCase:
		result := make([]dcbor.CBOR, 0, 1+len(c.Assertions))
		result = append(result, c.Subject.UntaggedCBOR())
		for _, assertion := range c.Assertions {
			result = append(result, assertion.UntaggedCBOR())
		}
		return dcbor.NewCBORArray(result)
	case *LeafCase:
		return dcbor.ToTaggedValue(dcbor.NewTag(bctags.TagLeaf, bctags.TagNameLeaf), *c.CBOR)
	case *WrappedCase:
		return c.Envelope.TaggedCBOR()
	case *AssertionCase:
		return c.Assertion.ToCBOR()
	case *ElidedCase:
		return dcbor.ToByteString(c.Digest.Bytes())
	case *KnownValueCase:
		return c.Value.UntaggedCBOR()
	case *EncryptedCase:
		return c.EncryptedMessage.TaggedCBOR()
	case *CompressedCase:
		return c.Compressed.TaggedCBOR()
	default:
		panic("unknown EnvelopeCase type")
	}
}

// TaggedCBOR returns the tagged CBOR representation of the envelope.
func (e *Envelope) TaggedCBOR() dcbor.CBOR {
	return dcbor.NewCBORTagged(envelopeCBORTags[0], e.UntaggedCBOR())
}

// ToCBOR returns the tagged CBOR representation.
func (e *Envelope) ToCBOR() dcbor.CBOR {
	return e.TaggedCBOR()
}

// ToCBORData returns the binary CBOR encoding of the envelope.
func (e *Envelope) ToCBORData() []byte {
	return e.TaggedCBOR().ToCBORData()
}

// EnvelopeFromTaggedCBORValue decodes a tagged CBOR value into an Envelope.
func EnvelopeFromTaggedCBORValue(cbor dcbor.CBOR) (*Envelope, error) {
	return dcbor.DecodeTagged(cbor, envelopeCBORTags, EnvelopeFromUntaggedCBOR)
}

// EnvelopeFromUntaggedCBOR decodes an untagged CBOR value into an Envelope.
func EnvelopeFromUntaggedCBOR(cbor dcbor.CBOR) (*Envelope, error) {
	cs := cbor.AsCase()
	switch cs.Kind {
	case dcbor.CBORKindTagged:
		tag, item, ok := cbor.AsTaggedValue()
		if !ok {
			return nil, fmt.Errorf("invalid tagged value in envelope")
		}
		tagVal := tag.Value()
		switch tagVal {
		case bctags.TagLeaf, bctags.TagEncodedCBOR:
			return newLeaf(item), nil
		case bctags.TagEnvelope:
			inner, err := EnvelopeFromTaggedCBORValue(cbor)
			if err != nil {
				return nil, err
			}
			return newWrapped(inner), nil
		case bctags.TagEncrypted:
			encrypted, err := bccomponents.DecodeEncryptedMessage(item)
			if err != nil {
				return nil, err
			}
			return newWithEncrypted(encrypted)
		case bctags.TagCompressed:
			compressed, err := bccomponents.DecodeCompressed(item)
			if err != nil {
				return nil, err
			}
			return newWithCompressed(&compressed)
		default:
			return nil, fmt.Errorf("unknown envelope tag: %d", tagVal)
		}

	case dcbor.CBORKindByteString:
		bs, err := cbor.TryIntoByteString()
		if err != nil {
			return nil, err
		}
		digest, err := bccomponents.DigestFromDataRef(bs)
		if err != nil {
			return nil, err
		}
		return newElided(digest), nil

	case dcbor.CBORKindArray:
		elements, err := cbor.TryIntoArray()
		if err != nil {
			return nil, err
		}
		if len(elements) < 2 {
			return nil, fmt.Errorf("node must have at least two elements")
		}
		subject, err := EnvelopeFromUntaggedCBOR(elements[0])
		if err != nil {
			return nil, err
		}
		assertions := make([]*Envelope, len(elements)-1)
		for i, elem := range elements[1:] {
			a, err := EnvelopeFromUntaggedCBOR(elem)
			if err != nil {
				return nil, err
			}
			assertions[i] = a
		}
		return newWithAssertions(subject, assertions)

	case dcbor.CBORKindMap:
		assertion, err := AssertionFromCBOR(cbor)
		if err != nil {
			return nil, err
		}
		return newWithAssertion(assertion), nil

	case dcbor.CBORKindUnsigned:
		value, _ := cbor.AsUnsigned()
		kv := knownvalues.NewKnownValue(value)
		return newWithKnownValue(kv), nil

	default:
		return nil, fmt.Errorf("invalid envelope")
	}
}

File: python/bc-envelope/src/bc_envelope/_cbor.py
"""CBOR encoding and decoding for Envelope.

The Gordian Envelope format uses CBOR with tag 200 (``TAG_ENVELOPE``).
Each envelope case has a unique CBOR signature enabling unambiguous
round-trip serialization.
"""

from __future__ import annotations

from bc_components import Compressed, Digest, EncryptedMessage
from bc_tags import (
    TAG_COMPRESSED,
    TAG_ENCODED_CBOR,
    TAG_ENCRYPTED,
    TAG_ENVELOPE,
    TAG_LEAF,
    tags_for_values,
)
from dcbor import CBOR, CBORCase, Map, Tag

from ._assertion import Assertion
from ._envelope import Envelope
from ._envelope_case import CaseType
from ._error import InvalidFormat
from known_values import KnownValue


# ---------------------------------------------------------------------------
# Tag constants
# ---------------------------------------------------------------------------

def envelope_tags() -> list[Tag]:
    """The CBOR tags identifying an Envelope."""
    return tags_for_values([TAG_ENVELOPE])


# ---------------------------------------------------------------------------
# Encoding
# ---------------------------------------------------------------------------

def untagged_cbor(envelope: Envelope) -> CBOR:
    """Encode *envelope* as untagged CBOR (the content inside the envelope tag)."""
    ct = envelope.case_type
    case = envelope.case

    if ct == CaseType.NODE:
        items: list[CBOR] = [untagged_cbor(case.subject)]
        for a in case.assertions:
            items.append(untagged_cbor(a))
        return CBOR.from_array(items)

    if ct == CaseType.LEAF:
        return CBOR.from_tagged_value(TAG_LEAF, case.cbor)

    if ct == CaseType.WRAPPED:
        return tagged_cbor(case.envelope)

    if ct == CaseType.ASSERTION:
        return case.assertion.to_cbor()

    if ct == CaseType.ELIDED:
        return case.digest.untagged_cbor()

    if ct == CaseType.KNOWN_VALUE:
        return case.value.untagged_cbor()

    if ct == CaseType.ENCRYPTED:
        return case.encrypted_message.tagged_cbor()

    if ct == CaseType.COMPRESSED:
        return case.compressed.tagged_cbor()

    raise AssertionError(f"unknown case type: {ct}")  # pragma: no cover


def tagged_cbor(envelope: Envelope) -> CBOR:
    """Encode *envelope* as tagged CBOR (``#6.200(...)``.)"""
    tags = envelope_tags()
    return CBOR.from_tagged_value(tags[0], untagged_cbor(envelope))


def cbor_data(envelope: Envelope) -> bytes:
    """Encode *envelope* to raw CBOR bytes."""
    return tagged_cbor(envelope).to_cbor_data()


# ---------------------------------------------------------------------------
# Decoding
# ---------------------------------------------------------------------------

def from_tagged_cbor(cbor: CBOR) -> Envelope:
    """Decode an Envelope from tagged CBOR (expects ``#6.200(...)``.)"""
    item = cbor.try_expected_tagged_value(envelope_tags()[0])
    return from_untagged_cbor(item)


def from_untagged_cbor(cbor: CBOR) -> Envelope:
    """Decode an Envelope from untagged CBOR content."""
    case = cbor.case

    # Tagged values
    if case == CBORCase.TAGGED:
        tag, item = cbor.try_tagged_value()
        tag_val = tag.value

        if tag_val in (TAG_LEAF, TAG_ENCODED_CBOR):
            return Envelope.new_leaf(item)

        if tag_val == TAG_ENVELOPE:
            inner = from_tagged_cbor(cbor)
            return Envelope.new_wrapped(inner)

        if tag_val == TAG_ENCRYPTED:
            encrypted = EncryptedMessage.from_untagged_cbor(item)
            return Envelope.new_with_encrypted(encrypted)

        if tag_val == TAG_COMPRESSED:
            compressed = Compressed.from_untagged_cbor(item)
            return Envelope.new_with_compressed(compressed)

        raise InvalidFormat(f"unknown envelope tag: {tag_val}")

    # Byte string → elided (32-byte digest)
    if case == CBORCase.BYTE_STRING:
        data = cbor.try_byte_string()
        return Envelope.new_elided(Digest.from_data(data))

    # Array → node (subject + assertions)
    if case == CBORCase.ARRAY:
        elements = cbor.try_array()
        if len(elements) < 2:
            raise InvalidFormat("node must have at least two elements")
        subject = from_untagged_cbor(elements[0])
        assertions = [from_untagged_cbor(e) for e in elements[1:]]
        return Envelope.new_with_assertions(subject, assertions)

    # Map → assertion
    if case == CBORCase.MAP:
        return Envelope.new_with_assertion(Assertion.from_cbor(cbor))

    # Unsigned integer → known value
    if case == CBORCase.UNSIGNED:
        kv = KnownValue(cbor.try_int())
        return Envelope.new_with_known_value(kv)

    raise InvalidFormat("invalid envelope")


def from_cbor_data(data: bytes | bytearray) -> Envelope:
    """Decode an Envelope from raw CBOR bytes."""
    cbor = CBOR.from_data(data)
    return from_tagged_cbor(cbor)


# ---------------------------------------------------------------------------
# Monkey-patch Envelope with CBOR methods
# ---------------------------------------------------------------------------

def _env_untagged_cbor(self: Envelope) -> CBOR:
    return untagged_cbor(self)


def _env_tagged_cbor(self: Envelope) -> CBOR:
    return tagged_cbor(self)


def _env_cbor_data(self: Envelope) -> bytes:
    return cbor_data(self)


Envelope.untagged_cbor = _env_untagged_cbor  # type: ignore[attr-defined]
Envelope.tagged_cbor = _env_tagged_cbor  # type: ignore[attr-defined]
Envelope.cbor_data = _env_cbor_data  # type: ignore[attr-defined]
Envelope.from_tagged_cbor = staticmethod(from_tagged_cbor)  # type: ignore[attr-defined]
Envelope.from_untagged_cbor = staticmethod(from_untagged_cbor)  # type: ignore[attr-defined]
Envelope.from_cbor_data = staticmethod(from_cbor_data)  # type: ignore[attr-defined]
